File: api/app/modules/eoq/solver.py
from __future__ import annotations

import logging
import math

from app.modules.eoq.models import EOQRequest
from app.schemas.common import SolveStatus
from app.schemas.result import GraphXY, ModuleResult, NamedTable, SolutionBlock

logger = logging.getLogger(__name__)


def solve(req: EOQRequest) -> ModuleResult:
    q_star = math.sqrt(2.0 * req.D * req.S / req.H)
    orders_per_year = req.D / q_star
    time_between = 1.0 / orders_per_year
    tc_ordering = orders_per_year * req.S
    tc_holding = (q_star / 2.0) * req.H
    tc = tc_ordering + tc_holding + req.C * req.D

    q_min = 0.2 * q_star
    q_max = 2.5 * q_star
    n = req.graph_points
    xs: list[float] = []
    tc_ys: list[float] = []
    ord_ys: list[float] = []
    hold_ys: list[float] = []
    for i in range(n):
        q = q_min + (q_max - q_min) * i / (n - 1)
        xs.append(q)
        n_orders = req.D / q
        ordering = n_orders * req.S
        holding = (q / 2.0) * req.H
        ord_ys.append(ordering)
        hold_ys.append(holding)
        tc_ys.append(ordering + holding + req.C * req.D)

    graph = GraphXY(
        type="xy",
        series=[
            {"name": "TC", "x": xs, "y": tc_ys},
            {"name": "ordering", "x": xs, "y": ord_ys},
            {"name": "holding", "x": xs, "y": hold_ys},
        ],
        x_label="Cantidad de pedido (Q)",
        y_label="Costo anual",
        title="Costos de inventario vs. cantidad de pedido",
        subtitle=f"Mínimo en Q* = {q_star:.2f}",
    )

    table = NamedTable(
        name="summary",
        columns=["métrica", "valor"],
        rows=[
            ["Q_star", q_star],
            ["orders_per_year", orders_per_year],
            ["time_between_orders_years", time_between],
            ["TC", tc],
            ["TC_ordering", tc_ordering],
            ["TC_holding", tc_holding],
        ],
    )

    result = ModuleResult(
        module="eoq",
        status=SolveStatus.ok,
        solution=SolutionBlock(
            variables={"Q": q_star},
            metrics={
                "Q_star": q_star,
                "orders_per_year": orders_per_year,
                "time_between_orders_years": time_between,
                "TC": tc,
                "TC_ordering": tc_ordering,
                "TC_holding": tc_holding,
            },
        ),
        iterations=None,
        sensitivity=None,
        graph=graph,
        tables=[table],
        warnings=[],
    )
    logger.info("module=%s status=%s", result.module, result.status.value)
    return result

File: worker/src/modules/eoq/solver.ts
import { SolverError } from "../../errors";
import { LIMITS, assertLimit } from "../../limits";
import type { GraphXY, ModuleResult } from "../../schema";
import { okResult } from "../../schema";

export type EoqRequest = {
  D: number;
  S: number;
  H: number;
  C: number;
  graph_points: number;
};

function asRecord(body: unknown): Record<string, unknown> {
  if (!body || typeof body !== "object" || Array.isArray(body)) {
    throw new SolverError("payload debe ser un objeto JSON");
  }
  return body as Record<string, unknown>;
}

function pos(v: unknown, name: string): number {
  const n = Number(v);
  if (!Number.isFinite(n) || !(n > 0)) throw new SolverError(`${name} debe ser > 0`);
  return n;
}

function nonneg(v: unknown, name: string, fallback: number): number {
  if (v == null || v === "") return fallback;
  const n = Number(v);
  if (!Number.isFinite(n) || n < 0) throw new SolverError(`${name} debe ser ≥ 0`);
  return n;
}

export function parseEoqRequest(body: unknown): EoqRequest {
  const o = asRecord(body);
  const graphPoints = o.graph_points == null ? 40 : Number(o.graph_points);
  if (!Number.isFinite(graphPoints) || graphPoints < 5 || graphPoints > LIMITS.eoqGraphPoints) {
    throw new SolverError(`graph_points debe estar entre 5 y ${LIMITS.eoqGraphPoints}`);
  }
  assertLimit(graphPoints <= LIMITS.eoqGraphPoints, `graph_points no puede superar ${LIMITS.eoqGraphPoints}`);
  return {
    D: pos(o.D, "D"),
    S: pos(o.S, "S"),
    H: pos(o.H, "H"),
    C: nonneg(o.C, "C", 0),
    graph_points: Math.trunc(graphPoints),
  };
}

export function solve(body: unknown): ModuleResult {
  const req = parseEoqRequest(body);
  const qStar = Math.sqrt((2.0 * req.D * req.S) / req.H);
  const ordersPerYear = req.D / qStar;
  const timeBetween = 1.0 / ordersPerYear;
  const tcOrdering = ordersPerYear * req.S;
  const tcHolding = (qStar / 2.0) * req.H;
  const tc = tcOrdering + tcHolding + req.C * req.D;

  const qMin = 0.2 * qStar;
  const qMax = 2.5 * qStar;
  const n = req.graph_points;
  const xs: number[] = [];
  const tcYs: number[] = [];
  const ordYs: number[] = [];
  const holdYs: number[] = [];
  for (let i = 0; i < n; i++) {
    const q = qMin + ((qMax - qMin) * i) / (n - 1);
    xs.push(q);
    const nOrders = req.D / q;
    const ordering = nOrders * req.S;
    const holding = (q / 2.0) * req.H;
    ordYs.push(ordering);
    holdYs.push(holding);
    tcYs.push(ordering + holding + req.C * req.D);
  }

  const graph: GraphXY = {
    type: "xy",
    series: [
      { name: "TC", x: xs, y: tcYs },
      { name: "ordering", x: xs, y: ordYs },
      { name: "holding", x: xs, y: holdYs },
    ],
    x_label: "Cantidad de pedido (Q)",
    y_label: "Costo anual",
    title: "Costos de inventario vs. cantidad de pedido",
    subtitle: `Mínimo en Q* = ${qStar.toFixed(2)}`,
  };

  return okResult("eoq", {
    status: "ok",
    variables: { Q: qStar },
    metrics: {
      Q_star: qStar,
      orders_per_year: ordersPerYear,
      time_between_orders_years: timeBetween,
      TC: tc,
      TC_ordering: tcOrdering,
      TC_holding: tcHolding,
    },
    graph,
    tables: [
      {
        name: "summary",
        columns: ["métrica", "valor"],
        rows: [
          ["Q_star", qStar],
          ["orders_per_year", ordersPerYear],
          ["time_between_orders_years", timeBetween],
          ["TC", tc],
          ["TC_ordering", tcOrdering],
          ["TC_holding", tcHolding],
        ],
      },
    ],
  });
}
